### DISCUSSION TITLE
Unable to write to log file

### DISCUSSION THREAD

**Discussion by ghost** [dani-garcia/vaultwarden#3535] [Q&A]
Hi, 

I have an issue with the latest version of vaultwarden. I installed it on LXD container without problems but when i keep the LOG FILE parameter in the .env file, it's impossible to start the service. I need to comment this line to be able to start vaultwarden.
Here is the error in the service log
`
root@VAULTWARDEN:/var/log# systemctl status vaultwarden
* vaultwarden.service - Vaultwarden Server (Rust Edition)
     Loaded: loaded (/etc/systemd/system/vaultwarden.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Mon 2023-05-22 15:32:00 UTC; 4s ago
       Docs: https://github.com/dani-garcia/vaultwarden
    Process: 785 ExecStart=/opt/vaultwarden/vaultwarden (code=exited, status=12)
   Main PID: 785 (code=exited, status=12)
        CPU: 8ms

May 22 15:32:00 VAULTWARDEN vaultwarden[785]: |   https://github.com/dani-garcia/vaultwarden/discussions or        |
May 22 15:32:00 VAULTWARDEN vaultwarden[785]: |   https://vaultwarden.discourse.group/                             |
May 22 15:32:00 VAULTWARDEN vaultwarden[785]: | Report suspected bugs/issues in the software itself at:            |
May 22 15:32:00 VAULTWARDEN vaultwarden[785]: |   https://github.com/dani-garcia/vaultwarden/issues/new            |
May 22 15:32:00 VAULTWARDEN vaultwarden[785]: \--------------------------------------------------------------------/
May 22 15:32:00 VAULTWARDEN vaultwarden[785]: [INFO] Using environment file `.env` for configuration.
May 22 15:32:00 VAULTWARDEN vaultwarden[785]: Error loading config:
May 22 15:32:00 VAULTWARDEN vaultwarden[785]:   Unable to write to log file. /var/log/vaultwarden.log
May 22 15:32:00 VAULTWARDEN systemd[1]: vaultwarden.service: Main process exited, code=exited, status=12/n/a
May 22 15:32:00 VAULTWARDEN systemd[1]: vaultwarden.service: Failed with result 'exit-code'.
root@VAULTWARDEN:/var/log#
`

i tried to create manually vaultwarden.log in /var/log and set the good permissions but it doesn't fix it. Actually i comment this line in .env LOG_FILE=/var/log/vaultwarden.log to launch the service. If anyone had already this problem please tell me how to bypass. Thanks :)

**Comment by ghost** [comment]
A friend gave me another instruction to add in the service file.

ReadWritePaths=/var/log/vaultwarden.log.

**Comment by sf-000** [comment]
Hello here,
maybe interassant for some guys.....

you have to add `ReadWritePaths=-/var/log/vaultwarden.log` if needed aditionally to @ghosts post on vaultwarden.service file and restart.

your vaultwarden service starts hardend? Look at your vaultwarden.service file. On dietpi debian triexie, here:

#### Part 1 - Adding write permission for log-file
File: /usr/lib/systemd/system/vaultwarden.service 
```
# Hardening
PrivateTmp=true
PrivateDevices=true
ProtectHome=true
ProtectSystem=strict
ReadWritePaths=-/mnt/dietpi_userdata/vaultwarden
ReadWritePaths=-/var/log/vaultwarden.log                 <- add this line or another path where your log-file is
```
#### Part 2 - restart vaultwarden
`systemctl restart vaultwarden`

#### Part 3 - Check logs
`journalctl -u vaultwarden.service -n20
`

**Comment by sf-000** [comment]
Have find a better solution for persistent fix:

## Update-safe fix: Unable to write to log file

**Environment:** DietPi / Debian (also applies to any distro with a hardened vaultwarden service)

### Problem

After a package update, vaultwarden fails to write logs:
```
Unable to write to log file. /var/log/vaultwarden.log
```

This happens because the service runs with systemd hardening enabled (`ProtectSystem=strict`), which blocks writes outside of explicitly allowed paths. The common fix is to add `ReadWritePaths=-/var/log/vaultwarden.log` to the service file — but editing `/usr/lib/systemd/system/vaultwarden.service` directly means **your change gets wiped on every package update**.

### Solution: systemd drop-in override

Instead of touching the package-managed file, create an override that persists across updates:

```bash
sudo systemctl edit vaultwarden
```

This automatically creates and opens:
```
/etc/systemd/system/vaultwarden.service.d/override.conf
```

Add only the line you need — no need to copy the full service file:

```ini
[Service]
ReadWritePaths=-/var/log/vaultwarden.log
```

Save, then reload and restart:
```bash
sudo systemctl daemon-reload && sudo systemctl restart vaultwarden
```

Confirm no more errors:
```bash
journalctl -u vaultwarden.service -n20
```

You can also verify the merged configuration (original + override) with:
```bash
systemctl cat vaultwarden
```

### Why this works

| Path | Managed by |
|------|------------|
| `/usr/lib/systemd/system/vaultwarden.service` | Package manager — overwritten on updates |
| `/etc/systemd/system/vaultwarden.service.d/override.conf` | You — persists across updates |

systemd merges both files automatically, so your override only needs to contain the lines you want to add or change.

> **Tip:** The `-` prefix in `ReadWritePaths=-/path` is intentional — it tells systemd to ignore the entry if the path doesn't exist yet, preventing startup failures on a fresh install.
